### DISCUSSION TITLE
Generating software bill of materials at component level

### DISCUSSION THREAD

**Discussion by PragatiGarg-eaton** [zephyrproject-rtos/zephyr#69755] [Q&A]
I am exploring the SBOM generation feature present in the west toolchain.

- I can generate the Software bill of materials in spdx format at source file level. However, to import the same in the BlackDuck, it doesn't accept the generated output file from West, which says corrupted file. 

![image](https://github.com/zephyrproject-rtos/zephyr/assets/107679628/ad69a7e7-7941-4a30-8a9b-4e42edd2889a)

- After digging deeper, I got the understanding the Blackduck accepts the sbom in a specific format, understands at component level and might not understand at the source level. 

zephr.spdx file generated by west spdx: [zephyr.spdx.json](https://github.com/zephyrproject-rtos/zephyr/files/14481596/zephyr.spdx.json)

Please suggest if there is any way to generate SBOM at the component level instead of at the source level.
 
Commands used to build:
build command: `west spdx -d BUILD_DIR`
Question: How can we generate sbom at component level instead of at source level using west toolchain?

OS: Windows

Links: https://docs.zephyrproject.org/latest/develop/west/zephyr-cmds.html#software-bill-of-materials-west-spdx
https://www.zephyrproject.org/how-to-generate-a-software-bill-of-materials-sbom-with-open-source-standards-and-tooling/
https://www.zephyrproject.org/generating-sboms-for-iot-at-build-time/

**Comment by kartben** [comment]
@tgagneret-embedded might be able to provide some insights

**Comment by yashi** [comment]
> zephr.spdx file generated by west spdx: [zephyr.spdx.json](https://github.com/zephyrproject-rtos/zephyr/files/14481596/zephyr.spdx.json)

I don't know what Black Duck SPDX does, but as you can see from the file above the file format West generate is not JSON.

There seems to be online converters if you search the net.

**Comment by tgagneret-embedded** [comment]
Hi @PragatiGarg-eaton ,

So if I understand you correctly you want the source files NOT to be present in the SPDX is that right ?
If so, I could add an option to `west spdx` to remove the source file from the output SBOM, it shouldn't be an issue (even though I need to confirm this).

For your information, I don't think the current SBOM will be useable by external tools right now. There is one PR waiting to be merged (#66495) and then modules would need to add information too.

**Comment by PragatiGarg-eaton** [comment]
> So if I understand you correctly you want the source files NOT to be present in the SPDX is that right? 

I am not necessarily saying that the source file should not be present in the SBOM. I am more interested in getting the SBOM at the module/component level. And if I could understand your PR correctly, it will be handling the same, right?

**Comment by tgagneret-embedded** [comment]
Looking at the `zephr.spdx(.json)`, I understand what you are saying. 

What you describe has been added to 3.6.0 (related PR: https://github.com/zephyrproject-rtos/zephyr/pull/66182). Did you use Zephyr 3.6.0 to generate your SBOM ?

**Reply by PragatiGarg-eaton** [reply]
No, I have been using zephyr 3.2.0, let me try it out with 3.6.0 once.
